src/backend/cuda/cast.hpp

```diff
@@ -89,9 +89,40 @@ struct CastWrapper {
     Array<To> operator()(const Array<Ti> &in) {
         CastOp<To, Ti> cop;
         common::Node_ptr in_node = in.getNode();
-        common::UnaryNode *node  = new common::UnaryNode(
-            static_cast<af::dtype>(dtype_traits<To>::af_type), cop.name(),
-            in_node, af_cast_t);
+        af::dtype to_dtype = static_cast<af::dtype>(dtype_traits<To>::af_type);
+
+        // JIT optimization in the cast of multiple sequential casts that become
+        // idempotent - check to see if the previous operation was also a cast
+        // TODO: handle arbitrarily long chains of casts
+        auto in_node_nary =
+            std::dynamic_pointer_cast<common::NaryNode>(in_node);
+        if (in_node_nary && in_node_nary->getOp() == af_cast_t) {
+            // The only way to get the input type of the child node if it's a
+            // cast is to get the output type of the child's child.
+            auto in_node_children = in_node_nary->getChildren();
+            // Check if any children are casts with the same type - if so,
+            // insert a shortcut noop
+            for (size_t i = 0; i < in_node_children.size(); ++i) {
```

Comment:
use `NaryNode::m_num_childern instead of size here. The size function will always return `Node::kMaxChildren`. You may need to create an accessor function.

Reply:
Since cast is a unary operation, you cast to UnaryNode instead of NaryNode. That way you can remove the loop.